### DISCUSSION TITLE
Advanced Component Queries

### DISCUSSION THREAD

**Discussion by roo-oliv** [Doraku/DefaultEcs#190] [Q&A] [answered]
If I'm not missing something, the only way to query Entities is via the presence (or absence) of Components. I find this limitation makes my code a little too loose and harder to read. Let me explain.

Say I have a `Collidable` component to indicate which entities should be checked for collisions. Now, I don't want to check every collidable entity against each other as it would be inefficient but rather I just want to check active/moving entities against all collidable entities. To do that I would have to create a new component, say `CollisionActor`, to then use this in the system's query. And this can get more and more complicated, I could have a collision cool down period which I would have to implement via the temporary deletion of the `Collidable` component or via another component `IgnoreCollision`.

I think it would be nice to be able to do a component like:
```c-sharp
public struct Collidable
{
    public bool IsActor;
    public int GracePeriod;
    public int CollisionGroup;

    public Collidable(bool isActor = true, int gracePeriod = 0, int collisionGroup = 0)
    {
        IsActor = isActor;
        GracePeriod = gracePeriod;
        CollisionGroup = collisionGroup;
    }
}
```

And since there are no custom filtering lambdas I can pass over to DefaultECS I'd have to filter the entities myself:
```c-sharp
    protected override void Update(GameState state, ReadOnlySpan<Entity> entities)
    {
        foreach (var entity in entities)
        {
            if(!entity.IsActor || entity.gracePeriod > 0 || entity.collisionGroup == 2) continue;
            ...
```

But this "if" would cause a branch divergence as far as I know and I'd lose performance. Also, I realize that maybe it's impractical to implement a filtering functionality within DefaultECS since you'd give up on using spans... unless you manage to create a more complex system of declaring member properties of a component as indexes I think.

Am I better off just creating new Components for these "properties" so I can filter my entities or is there a better way?

**Comment by Doraku** [comment, marked as answer]
You can actually query on the component value too using the [With(ComponentPredicate)](https://github.com/Doraku/DefaultEcs/blob/master/documentation/api/EntityQueryBuilder.EitherBuilder.With_T_(ComponentPredicate_T_).md) api. Only thing to keep in mind when using it, you need to [Notify](https://github.com/Doraku/DefaultEcs/blob/master/documentation/api/Entity.NotifyChanged_T_().md) explicitly (doing an `entity.Set(newValue)` does it implicitely, using `entity.Get<T>() = newValue` does not) that the value has changed so the systems can recheck the entity.

**Reply by roo-oliv** [reply]
Oh, I missed this completely! Thanks for pointing it out!
